sha2/src/sha512/x86.rs

```diff
@@ -0,0 +1,354 @@
+//! SHA-512 `x86`/`x86_64` backend
+
+#![allow(clippy::many_single_char_names)]
+
+use core::mem::size_of;
+
+#[cfg(target_arch = "x86")]
+use core::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::*;
+
+use crate::consts::{K64, K64X4};
+
+cpufeatures::new!(avx2_cpuid, "avx", "avx2", "sse2", "sse3");
+
+pub fn compress(state: &mut [u64; 8], blocks: &[[u8; 128]]) {
+    // TODO: Replace with https://github.com/rust-lang/rfcs/pull/2725
+    // after stabilization
+    if avx2_cpuid::get() {
+        unsafe {
+            sha512_compress_x86_64_avx2(state, blocks);
+        }
+    } else {
+        super::soft::compress(state, blocks);
+    }
+}
+
+#[target_feature(enable = "avx,avx2,sse2,sse3")]
```

Comment:
AVX2 [implies](https://doc.rust-lang.org/reference/attributes/codegen.html#x86-or-x86_64) AVX. And this chain goes further: AVX2 -> AVX -> SSE4.2 -> SSE4.1 -> SSSE3 -> SSE3 -> SSE2. In other words, you only need to enable `avx2`.